Update README to be ergonomic and excite new users (#816)
datafold/data-diff: README.md

</h2>
<br>

> [Join our live virtual lab series to learn how to set it up!](https://www.datafold.com/virtual-hands-on-lab)

# What's a Data Diff?
A data diff is the value-level comparison between two tables—used to identify critical changes to your data and guarantee data quality.

There is a lot you can do with data-diff: you can test SQL code by comparing development or staging environment data to production, or compare source and target data to identify discrepancies when moving data between databases.

# Use Cases

### Data Migration & Replication Testing
data-diff is a powerful tool for comparing data when you're moving it between systems. Use it to ensure data accuracy and identify discrepancies during tasks like:

> - **Migrating** to a new data warehouse (e.g., Oracle > Snowflake)

Author: ```suggestion
- **Migrating** to a new data warehouse (e.g., Oracle -> Snowflake)
```

Reviewer: fixed

> - **Converting SQL** to a new transformation framework (e.g., stored procedures > dbt)

Author: ```suggestion
- **Converting SQL** to a new transformation framework (e.g., stored procedures -> dbt)
```

Reviewer: fixed

> - Continuously **replicating data** from an OLTP database to OLAP data warehouse (e.g., MySQL > Redshift)

Author: ```suggestion
- Continuously **replicating data** from an OLTP database to OLAP data warehouse (e.g., MySQL -> Redshift)
```

Reviewer: fixed

2. Run the SQL code to create a new dataset
3. Compare this dataset with its production version or other iterations

# dbt Integration
  <p align="left">
  <img alt="dbt" src="https://seeklogo.com/images/D/dbt-logo-E4B0ED72A2-seeklogo.com.png" width="10%" />
  </p> 

data-diff integrates with [dbt Core](https://github.com/dbt-labs/dbt-core) to seamlessly compare local development to production datasets.

Learn more about how data-diff works with dbt:
* Read our docs to get started with [data-diff & dbt](https://docs.datafold.com/development_testing/cli) or :eyes: **watch the [4-min demo video](https://www.loom.com/share/ad3df969ba6b4298939efb2fbcc14cde)**
* dbt Cloud users should check out [Datafold's out-of-the-box deployment testing integration](https://www.datafold.com/data-deployment-testing)
* Get support from the dbt Community Slack in [#tools-datafold](https://getdbt.slack.com/archives/C03D25A92UU)

# Getting Started

### ⚡ Validating dbt model changes between dev and prod
Looking to use data-diff in dbt development? 

Development testing with Datafold enables you to see the impact of dbt code changes on data as you write the code, whether in your IDE or CLI. 

>  Head over to [our `data-diff` + `dbt` documentation](https://docs.datafold.com/development_testing/how_it_works) to get started with a development testing workflow!

Author: ```suggestion
 Head over to [our `data-diff` + `dbt` documentation](https://docs.datafold.com/development_testing/cli/) to get started with a development testing workflow!
```

Reviewer: fixed

+ 1, returned
```

5. Review the output

After running your data-diff job, review the output to identify and analyze differences in your data.

Check out [documentation](https://docs.datafold.com/reference/open_source/cli) for the full command reference.

> # data-diff OSS & Datafold Cloud

Author: Let's move this section right above the contributions section. This doesn't narratively make sense to inject here.

Reviewer: fixed

> data-diff is an open source utility. You run stateless diffs on your local computer but results cannot be easily accessed by others.

Author: ```suggestion
data-diff is an open source utility for running stateless diffs on your local computer for a great single player experience.
```

<br>

# How it works

`data-diff` efficiently compares data using two modes:

> 1. **`joindiff`**: Ideal for comparing data within the same database, utilizing outer joins for efficient row comparisons. It relies on the database engine for computation and has consistent performance.

Author: Can you convert these into toggles for both joindiff and hashdiff? Let's not force readers to context switch for a quick explainer. I agree with moving the how it works section here. 

## `joindiff`
- Recommended for comparing data within the same database
- Uses the outer join operation to diff the rows as efficiently as possible within the same database
- Fully relies on the underlying database engine for computation
- Requires both datasets to be queryable with a single SQL query
- Time complexity approximates JOIN operation and is largely independent of the number of differences in the dataset

## `hashdiff`
- Recommended for comparing datasets across different databases
- Can also be helpful in diffing very large tables with few expected differences within the same database
- Employs a divide-and-conquer algorithm based on hashing and binary search
- Can diff data across distinct database engines, e.g., PostgreSQL <> Snowflake
- Time complexity approximates COUNT(*) operation when there are few differences
- Performance degrades when datasets have a large number of differences

Author: Keep your new summary sentences though.

Reviewer: That's a good option - did not think of toggles!

Reviewer: re-added
